src/model/consent/consent.ts

```diff
@@ -83,20 +83,44 @@ export class ConsentDB {
   }
 
   // Update Consent
-  // No validation against Null or illegal updates in models
+  // Only validating against updating existing fields
   public async update (consent: Consent): Promise<number> {
     // Returns number of updated rows
-    const updateCount: number = await this
-      .Db<Consent>('Consent')
-      .where({ id: consent.id })
-      .update(consent)
-
-    // Ensure that the caller knows that the resource does not exist
-    if (updateCount === 0) {
-      throw new NotFoundError('Consent', consent.id)
-    }
-
-    return updateCount
+    // Transaction to make the update atomic
+    return this.Db.transaction(async (trx): Promise<number> => {
+      // Transaction is rolled back automatically if there is
+      // an error and the returned promise is rejected
+      const consents: Consent[] = await trx<Consent>('Consent')
+        .select('*')
+        .where({ id: consent.id })
+        .limit(1)
+
+      if (consents.length === 0) {
+        throw new NotFoundError('Consent', consent.id)
+      }
+
+      const existingConsent = consents[0]
+      const updatedConsent: Consent = { ...consent }
+
+      // Cannot update fields once they have been added properly to the DB
+      for (const key in consent) {
+        // Only include the fields which are NULL in the DB
+        // Include credentialStatus even if NON-NULL but ensure that it is not 'ACTIVE'
+        if ((existingConsent[key as keyof Consent] !== null && key !== 'credentialStatus') ||
+            (key === 'credentialStatus' && existingConsent[key as keyof Consent] === 'ACTIVE')) {
+          delete updatedConsent[key as keyof Consent]
+        }
+      }
```

Comment:
There's a lot of hard to read logic in this for statement contained in the `if statement`.

My preference would be to use a JS forEach with a function to build up the `updatedFields` object, for example:

```ts
const updatedFields: { [index: string]: string } = {};
Object.keys(consent).forEach(key => {
  const value = existingConsent[key as keyof Consent]

  // Don't allow overwriting Credential status
  if (key === 'credentialStatus' && value === 'ACTIVE') {
    return;
  }

  // Don't allow overwriting non-null fields
  if (value !== null) {
    return;
  }

  // Ensure type safety - the only fields we wish to update should be strings
  if(typeof value !== 'string') {
    return;
  }

  updatedFields[key] = value;
})
```

> (I don't know if I got the logic right, but hopefully you get the idea)

Reply:
@lewisdaly Isn't `forEach` slower than a `for loop` and the older way to do this? Also, `updatedFields` would be of type `Object` instead of `Consent` and we would have to typecast it to `Consent`.

Reply:
Also, the type validation is provided by TS and will be a responsibility of the handler.

Reply:
@ramanmangla forEach is slower, but the real difference will be visible for more than hundreds of items on the list. @lewisdaly 's code is much cleaner! I would suggest to use it. Type validation is only done at compile-time, at execution -time all evils are waiting for your soul.

Reply:
Indeed I agree with @eoln's agreeing to me. By passing in a function that gets evaluated over every key of the object, we get nice and easy ways to `return` early in case we don't want do anything, which is much easier to read than `deleting` in my opinion.

Reply:
is it called mutual agreement @lewisdaly

Reply:
but for loop, you can use the `continue` statement in the same you are using `return` in callback

Reply:
I definitely agree that the code would be much cleaner. I will make the changes.